### PR TITLE
Combine all Circuit UI packages into a monorepo

### LINKED ISSUE #703: [RFC] Monorepo
## Suggested technology

Circuit UI has grown from a single package (`@sumup/circuit-ui`) into an **ecosystem** of packages ([`@sumup/design-tokens`](https://www.npmjs.com/package/@sumup/design-tokens), [`@sumup/icons`](https://www.npmjs.com/package/@sumup/icons)). Each package currently lives in its own repository. In the review process for open-sourcing the new subpackages, @danpersa proposed to combine all packages in a single **monorepo**.

This would offer several benefits:

- **Community management**: Issues, bug reports, and feature requests would be filed in one place where they can be easily searched and triaged. 
- **Coordinated releases**: Sometimes, a task requires changes in multiple packages. In a monorepo, related changes can be done in one commit and one PR which simplifies code review and the coordination of the release.
- **Centralized documentation**: Having a single place for documentation of all packages makes it easier for users to learn about the ecosystem and how the packages relate to each other. [`@sumup/icons`](https://www.npmjs.com/package/@sumup/icons) is already documented in the Storybook of the component library, however, the list of icons must be kept up-to-date manually.

### Folder Structure

The packages would live in a folder aptly called `packages`. The Storybook configuration, general documentation, and shared linting configuration files would live on the top level of the repository.

```sh
├── .storybook/
├── docs/
├── packages/
│   ├── cli/
│   ├── circuit-ui/
│   ├── design-tokens/
│   └── icons/
├── prettier.config.js
├── README.md
└── ...
```

Each package would contain its source code, testing & build configs, and a separate README.

```sh
components/
├── src/
├── jest.config.json
├── tsconfig.json
├── README.md
└── ...
```

### Tooling

Monorepos add some complexity when it comes to shared tooling and dependency management. These tools are here to help:

#### [yarn workspaces](https://classic.yarnpkg.com/en/docs/workspaces)

`yarn`, our package manager of choice, has built-in support for monorepos. Each package is a "workspace" and `yarn` can deduplicate dependencies between them. More importantly, packages can depend on _each other_ and always use the latest version.

#### [lerna](https://lerna.js.org/)

Lerna builds on top of `yarn workspaces` to provide additional tools and commands to manage a monorepo, e.g. running a package script (e.g. `yarn test`) in each package. It supports publishing as well, but doesn't handle release notes, which is why I propose to use another tool for that (see below 👇).

### Release Process

We currently use [`semantic-release`](https://github.com/semantic-release/semantic-release/) to automate our publishing and release note generation. `semantic-release` assumes a one-to-one relationship between a Git repository and NPM package. It relies on [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) to work, but in a monorepo, it would be unable to determine which commit affects which package. 

#### [`changesets`](https://github.com/atlassian/changesets)

I propose to use Atlassian's `changesets` tool instead. It enables the same release automation as `semantic-release` but with an explicit focus on monorepos. Some of its differences are:

- **No enforced commit message format**: While `semantic-release` requires [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) to infer the version type and changelog from the commit history, `changesets` creates the changelog based on files that are checked into git. 
- **More control over releases**: `semantic-release` publishes a new release for (almost) every commit, which can flood the release log. `changesets` publishes a new release only when it's asked to, allowing us to group changes together.

## Alternatives

The alternative would be to keep one repository for each package. As the ecosystem grows, maintenance and community management would become more difficult.

## Potential downsides

Monorepos come with additional complexity and require additional tooling to work well. I believe the initial investment pays off over the long run.

It becomes harder to tell issues/PRs/changelogs/etc apart. Which one refers to which package?

As the repository and its git history grow, it takes longer to clone it and set it up locally, which could discourage contributions.

### PR DESCRIPTION
**Pull request by connor-baer**
Closes #703.

## Purpose

Circuit UI has grown from a single package (`@sumup/circuit-ui`) into an **ecosystem** of packages ([`@sumup/design-tokens`](https://www.npmjs.com/package/@sumup/design-tokens), [`@sumup/icons`](https://www.npmjs.com/package/@sumup/icons)), [`create-sumup-react-app`](https://github.com/sumup/create-sumup-react-app), and [`create-sumup-next-app`](https://github.com/sumup/create-sumup-next-app). Each package currently lives in its own repository. Combining them into a single monorepo simplifies the **community management**, **documentation**, and **release process**. See #703 for the full reasoning.

## Approach and changes

- Set up [yarn workspaces](https://classic.yarnpkg.com/en/docs/workspaces) and [lerna](https://lerna.js.org/)
- Move the `@sumup/circuit-ui` code into the `/packages` folder
- Configure Jest and Eslint to work at the root as well as the package level
- Replace [semantic-release](https://github.com/semantic-release/semantic-release) with [changesets](https://github.com/atlassian/changesets) to automate the release process
- Update the README

---

This is the first step in the monorepo re-organization. The next PRs will cover:

- Move [design-tokens](github.com/sumup-oss/design-tokens), [icons](https://github.com/sumup-oss/icons), [create-sumup-react-app](https://github.com/sumup/create-sumup-react-app), and [create-sumup-next-app](https://github.com/sumup/create-sumup-next-app) into the monorepo
- Update the documentation for the release process to cover changesets

## Definition of done

* [x] Development completed
* [x] Reviewers assigned
* [x] Unit and integration tests
* [x] Meets minimum browser support
* [x] Meets accessibility requirements

### CODE REVIEW

**Review by amelako — APPROVED** [review]
🚀🥳 like it a lot!

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-name: Continous Integration
+name: Continuos Integration
```
Comment (robinmetral) [inline .github/workflows/release.yml]: Continuous

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -30,15 +30,15 @@ jobs:
...
       - name: Codecov
         uses: codecov/codecov-action@v1.0.3
         with:
           token: ${{ secrets.CODECOV_TOKEN }}
-      - name: Release
-        env:
-          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          NPM_TOKEN: ${{ secrets.NPM_TOKEN }}
-        run: yarn release
+      # - name: Release
```
Comment (robinmetral) [inline .github/workflows/release.yml]: Should we remove this?

File: README.md [hunk 3]
```diff
@@ -32,7 +33,7 @@ If you feel another member of the community violated our CoC or you are experien
 ### Maintainers
 
 - [Connor Bär](mailto:connor.baer@sumup.com)
-- [Fernando Fleury](mailto:fernando.fleury@sumup.com)
+- [Amela Kodzic](mailto:amela.kodzic@sumup.com)
```
Comment (robinmetral) [inline README.md:36]: ❤️

**Comment by robinmetral** [comment]
☝️ oops, reviewed commit by commit so most things were outdated 🙈 Looks great! I'm not familiar with things like changeset config, let's try it and tweak it as we go if necessary

**Pull request merged**